python/mlx/nn/layers/flatten.py

```diff
@@ -0,0 +1,90 @@
+import mlx.core as mx
+from mlx.nn.layers.base import Module
+
+
+def _get_new_shape(a: mx.array, start: int, end: int):
+    shape = a.shape
+
+    if a.ndim == 0 or (end == start):
+        return shape
+    end = min(end, a.ndim - 1)
+    length = shape[end]
+    for idx in range(end - 1, start - 1, -1):
+        if shape[idx] == 0 or shape[idx + 1] == 0:
+            length = 0
+            break
+
+        if shape[idx] == 1:
+            continue
+
+        length = length * shape[idx]
+
+    new_shape = (*shape[:start], length, *shape[end + 1 :])
+    return new_shape
+
+
+def flatten(a: mx.array, start_dim: int = 0, end_dim: int = -1) -> mx.array:
+    r"""
+    Flattens a contiguous range of dims into a tensor.
+
+    See :class:`~nn.Flatten` for details.
+
+    Args:
+        input: input tensor.
+        start_dim: first dim to flatten (default = 0).
+        end_dim: last dim to flatten (default = -1).
+    """
+
+    start_dim = max(0, start_dim)
+    end_dim = a.ndim if end_dim == -1 else min(a.ndim, end_dim)
+
+    if start_dim > end_dim:
+        raise ValueError("start_dim must be less than or equal to end_dim")
+
+    if start_dim == end_dim and a.ndim != 0:
+        return a
+    return mx.reshape(a, _get_new_shape(a, start_dim, end_dim))
```

Comment:
This should really be an op in `mx.core` (and follow the NumPy API please). Want to take a stab at moving it there and adding a Python binding? That would be awesome

Reply:
sure, I'm working on it! thanks for your input :p

Reply:
@awni can you please review up to this point?